### ISSUE TITLE
Tuning Parameters and metrics questions

### ISSUE THREAD

**Issue by skeller047** [go-graphite/go-carbon#15]
We're testing go-carbon as a possible replacement for python carbon.  Couple of questions have come up:

1) Do we need to ensure that persister.committedPoints is the same as pickle.metricsReceived (we're feeding go-carbon from a relay...).  I notice that the cache size immediately goes up to max if we keep IO down by setting max-updates-per-second.  If committedPoints is less than metricsReceived, are we losing data?

2) Is it possible to run multiple go-carbon instances on a single server?  Is this even desirable?

Finally, any guidance you can give on tuning - we have set max-updates-per-second, max-cpu and cache max-size to various values during testing, but it's not completely clear what the "best" way to set these is. 

Thanks!

**Comment by lomik** [comment]
1. Check cache.overflow graph. This is count of lost points.
2. That is unnecessary. All that possible paralleled in a single process.

We use this configuration on servers with many SSD drives:

```
[common]
max-updates-per-second = 0
max-cpu = <server cores - 1>

[whisper]
workers = <max-cpu * 2>

[cache]
max-size = <average points per minute * 10>
input-buffer = <average points per minute>
```

**Comment by skeller047** [comment]
Thanks a lot Roman I'll try those settings :)

**Comment by gerrickw** [comment]
We have 4 ssds in raid 0 and they seem very bored with the config settings provided above with 5-30 minute delays in metric write time.  In our case we have a write pattern of 630000 metrics happening during a 10 second window so possibly the average points per minute are skewed, based on queuing logic.   If I times that by 6 for a minute duration I get 3780000.  This causes the below config where we see writes to disks happening every 5-30 minutes, where we would of hoped close to 1 minute to 30 second delays.

[cache]
max-size = 37800000
input-buffer = 3780000

Using your formula above I got the following

[common]
max-updates-per-second = 0
max-cpu = 20

[whisper]
workers = 40

[cache]
max-size = 37800000
input-buffer = 3780000

I was messing a bit with max-size and input-buffer, but a bit confused how to tune the parameters in our case.  I am seeing no overflow in cache metrics or capping of iostat %util.  If there is a way to tune based on carbon metrics that would be useful as well.

Thanks for your help.

**Comment by gerrickw** [comment]
As a note, looking at writes.  We seem to see batching occurring.  Barely no IO activity for a while and then suddenly a huge number of writes.  Is there a way to make it write once metrics are received?

**Comment by skeller047** [comment]
I tried these settings on a system that uses iSCSI storage (platter driven).  I found that I needed to use the max-updates-per-second as a throttle to keep IOPS reasonable.  Since the storage is shared by many systems, the admins get a little annoyed if I run hotter than about 4k or 5k write IOPS.  This is similar to the way we tune python-carbon on the same type systems.

A way to gauge how good a job you are doing is if the cache size is reasonable (less than the max of course), and persister.pointsPerUpdate is > 1 and < 10.  This way, if you are sending 1 point per metric per minute, you have no more than 10 unwritten data points in cache at any time, but you keep IOPS down.

This is less of a concern with SSD based systems, as thrashing doesn't hurt them physically.
